docs/source/extending/overloading-guide.rst

```diff
@@ -0,0 +1,282 @@
+Providing multiple implementations and dispatching based on types
+=================================================================
+
+As you saw above, the overload implementation for ``set_to_x`` function
+doesn't accept floating-point arguments. Let's extended the specification of
+the function as follows:
+
+* The numerical type of the array ``arr`` must match the numerical type of the
+  scalar ``x`` argument, i.e. if ``arr`` is of type ``int64``, then ``x`` must
+  be of this type too.
+* Only integer and floating-point types are to be supported for argument ``x``.
+* No ``nan`` values are allowed in ``arr`` when it is of floating-point type and if
+  such a value is encountered an appropriate ``ValueError`` should be raised.
+* If a tuple is used instead of an array as a value for ``arr``, a custom error message with a hint
+  for the user should be issued.
+
+The resulting implementation could look as follows:
+
+.. literalinclude:: myjitmodule2.py
+
+As you can see, the typing checking code has been increased significantly to
+match the new requirements. Also, multiple implementations---one for integers
+and one for floating-point---are provided. We check inside the typing scope
+which implementation should be used and also raise any custom error messages
+required. Importantly, the check for ``nan`` values is only present in the
+floating point implementation as this additional check creates a runtime
+overhead. This can easily be observed during profiling:
+
+.. code:: pycon
+
+    In [1]: a = np.arange(100, dtype=np.float64)
+
+    In [2]: %timeit myalgorithm(a, 1.0)
+    473 ns ± 11.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
+
+    In [3]: a = np.arange(100)
+
+    In [4]: %timeit myalgorithm(a, 1)
+    237 ns ± 4.59 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
+
+Writing Tests
+=============
+
+The following is a sufficient piece of test code for the overloaded
+``set_to_x`` implementation. As you can see, only a small part of the test-code
+is about testing if the function works correctly. Most of the test code in this
+example checks that all error cases are handled and that all raised exceptions
+are of the correct type and have the correct error message. When implementing
+tests cases for Numba, you should always use ``numba.tests.support.TestCase``.
+
+.. literalinclude:: test_myjitmodule2.py
+
+While it is a pretty decent test, it wouldn't be accepted into Numba.
+There are a few more test-cases that should be implemented:
+
+* Empty array
+* Single value array
+* Multidimensional arrays
+* Tests for ``int32`` and ``float32`` types
+
+You can download the example code from above including tests here:
+
+* :download:`mymodule.py </extending/mymodule.py>`
+* :download:`myjitmodule.py </extending/myjitmodule.py>`
+* :download:`myjitmodule2.py </extending/myjitmodule2.py>`
+* :download:`test_myjitmodule2.py </extending/test_myjitmodule2.py>`
+
+
+Implementing ``@overload``s for NumPy functions
```

Comment:
This is causing a sphinx warning, ` WARNING: Inline literal start-string without end-string.`

Reply:
Yes, fixed, re-running CI now.